src/app/shell/item-comparators.ts

```diff
@@ -109,6 +110,114 @@ const TAG_ITEM_COMPARATORS: {
   archive: (getTag) => compareBy((item) => getTag(item) === 'archive'),
 };
 
+export type VaultGroupValue = string | number | boolean | undefined;
+
+interface MutableVaultGroup {
+  groupingValue: VaultGroupValue;
+  iconValue?: VaultGroupIconValue;
+  items: DimItem[];
+}
+
+interface VaultGroup {
+  groupingValue: VaultGroupValue;
+  iconValue?: VaultGroupIconValue;
+  items: readonly DimItem[];
+}
```

Comment:
There isn't really a need to enforce immutability via the type here, but if you want to, there's also `Draft<VaultGroup>` from the `immer` types.